### PR TITLE
Use ruamel.yaml for YAMLEditable

### PR DESCRIPTION
**Pull request by skowalak**
replace the YAMLEditable.to_yaml() method with ruamel serialization.
migrate some yaml formatting to the helpers module.

### CODE REVIEW

**Comment by skowalak** [comment]
This PR references (#276)
The entire `YAMLEditable.to_yaml()` function is rewritten to use `ruamel.yaml`

What I have not done (yet):
* Pretty printing still relies on `helpers.convert_to_yaml()`. I don't know if that level of indentation/alignment is possible with ruamel
* Values will be inserted in the standard template. I don't know if this breaks some workflows involving custom new user templates (#146).
* Helper tests still use `helpers.convert_to_yaml()`. I will add some tests for the new methods when I have some more time.

File: khard/helpers/__init__.py [hunk 1]
```diff
@@ -61,6 +62,101 @@ def get_random_uid() -> str:
...
+      2. list with only one item become this item
+      3. multiline strings use the YAML literal style:
+         https://yaml.org/spec/1.2.2/#literal-style
+    
+    :param value: the value to be sanitized
+    :returns: the sanitized value
+    """
+    # special case for empty values
+    if not value:
+        value = None
```
Comment (lucc) [inline khard/helpers/__init__.py:78]: You do not need the local variable `value` (or rather overwrite the argument), you can just `return` directly.

Also see https://pheanex.github.io/pylint/

File: khard/helpers/__init__.py [hunk 2]
```diff
@@ -61,6 +62,101 @@ def get_random_uid() -> str:
...
+    elif isinstance(value, str):
+        if "\n" in value:
+            value = LiteralScalarString(value)
+
+    return value
+
+
+def yaml_dicts(
+        data: Optional[Dict[str, Any]], 
+        defaults: Union[Dict[str, Any], List[str]] = None
```
Comment (lucc) [inline khard/helpers/__init__.py:92]: Shouldn't that be a `Optional[Union[...]]` if it can be `None?

File: khard/helpers/__init__.py [hunk 3]
```diff
@@ -61,6 +62,101 @@ def get_random_uid() -> str:
...
+    """
+    format a dict according to template, if empty use specified defaults
+
+    :param data: dict of contact data with keys as types and values as values.
+    :param defaults: default dict to be used if data is empty.
+    :returns: dict of types and values.
+    """
+    if not data:
+        if isinstance(defaults, list):
+            return dict((key, None) for key in defaults)
```
Comment (lucc) [inline khard/helpers/__init__.py:103]: You could use `dict.fromkeys` or at least  literal dict comprehension with `{ ... }` like below,  to keep your code style uniform.

File: khard/helpers/__init__.py [hunk 4]
```diff
@@ -61,6 +62,101 @@ def get_random_uid() -> str:
...
+       
+        return defaults
+
+    data_dict = {}
+    for t, nlist in sorted(data.items(), key=lambda k: k[0].lower()):
```
Comment (lucc) [inline khard/helpers/__init__.py:108]: Why t and nlist and not key, value?  What do these names stand for, I don't understand them?

File: khard/helpers/__init__.py [hunk 5]
```diff
@@ -61,6 +62,101 @@ def get_random_uid() -> str:
...
+    
+    return data_dict
+
+
+def yaml_addresses(addresses: Optional[Dict[str, Any]],
+                     address_properties: List[str],
+                     defaults: List[str] = []
+                    ) -> Dict[str, Any]:
+    if not addresses:
+        address_fields = dict((key, None) for key in address_properties)
```
Comment (lucc) [inline khard/helpers/__init__.py:119]: `dict.fromkeys`

File: khard/helpers/__init__.py [hunk 6]
```diff
@@ -61,6 +62,101 @@ def get_random_uid() -> str:
...
+    return address_dict
+
+
+def yaml_anniversary(anniversary: Union[str, datetime],
+                       version: str) -> Optional[str]:
+    if not anniversary:
+        return None
+
+    if isinstance(anniversary, str):
+        string = anniversary
```
Comment (lucc) [inline khard/helpers/__init__.py:138]: You can `return` instead of assigning to `string`.

File: khard/helpers/__init__.py [hunk 7]
```diff
@@ -61,6 +62,101 @@ def get_random_uid() -> str:
...
+                    ) -> Dict[str, Any]:
+    if not addresses:
+        address_fields = dict((key, None) for key in address_properties)
+        return {address_type: address_fields for address_type in defaults}
+
+    address_dict = {}
+    for address_type, address in addresses.items():
+        if isinstance(address, list):
+            address = address[0]
+        address_dict[address_type] = (
```
Comment (lucc) [inline khard/helpers/__init__.py:126]: outer parents are not needed.

File: khard/helpers/__init__.py [hunk 8]
```diff
@@ -61,6 +62,101 @@ def get_random_uid() -> str:
...
+        return defaults
+
+    data_dict = {}
+    for t, nlist in sorted(data.items(), key=lambda k: k[0].lower()):
+        data_dict[t] = nlist[0] if len(nlist) == 1 else nlist
+    
+    return data_dict
+
+
+def yaml_addresses(addresses: Optional[Dict[str, Any]],
```
Comment (lucc) [inline khard/helpers/__init__.py:114]: Please add a doc comment.

**Review by lucc — COMMENTED** [review]
I did a quick review of the helper code, just to give you an idea.  I will of course do another one at least when you remove the draft state.

**Comment by lucc** [comment]
I think you need not bother with #146 because that issue is still in the discussion and planing phase. You can also leave the pretty printing alone as we just reuse the yaml conversion functions.  We do not officially claim that the output of the pretty printer is valid yaml.  We can thing about changing the pretty printer in another PR.

Please try to run the tests  and mypy (and maybe even pylint) locally because currently  travis integration is broken. (I need help from @scheibler to fix this.)

If you could also write a test for #276 (the string with an @) that would be great!

**Comment by skowalak** [comment]
I have a little difficulty running the tests. When I run `python setup.py test` hangs at the line
```
test_contact_is_found_if_name_matches (test.test_command_line_interface.AddEmail) ... 
```
I have to interrupt the testing via `CTRL-C` because the process' memory usage grows beyond what I consider normal (10GiB). Is that correct? Could my changes have caused this behaviour, if incorrect?

**Comment by skowalak** [comment]
Hi @lucc ,
after some troubleshooting the abovementioned issue, I tried running the tests on the `develop` branch and had the exact same problem (`tests.test_command_line_interface.AddEmail` hangs in an infinite loop consuming more and more memory). I even tried it on a different computer running a different OS.

So apparently my changes did not break the tests.

- [x] I ran mypy
- [x] I ran pylint
- [x] I wrote a test for #276 
- [ ] I successfully ran the tests

Is there anything more I need to do before this PR can be merged?

File: khard/carddav_object.py [hunk 9]
```diff
@@ -1237,180 +1239,53 @@ def to_yaml(self) -> str:
...
-                    strings.append("Birthday : ")
-            elif line.lower().startswith("categories"):
-                strings += helpers.convert_to_yaml(
-                    "Categories", self.categories, 0, 11, True)
-            elif line.lower().startswith("note"):
-                strings += helpers.convert_to_yaml(
-                    "Note", self.notes, 0, 5, True)
-            elif line.lower().startswith("webpage"):
-                strings += helpers.convert_to_yaml(
-                    "Webpage", self.webpages, 0, 8, True)
```
Comment (lucc) [inline khard/carddav_object.py:1411]: I like this code reduction :)

**Review by lucc — CHANGES_REQUESTED** [review]
You can skip the broken test like this:
```diff
diff --git a/test/test_command_line_interface.py b/test/test_command_line_interface.py
index cac683c..97b5ecf 100644
--- a/test/test_command_line_interface.py
+++ b/test/test_command_line_interface.py
@@ -480,6 +480,7 @@ class Merge(unittest.TestCase):
         self.assertEqual('contact2.vcf', second)


+@unittest.skip("reason for skipping")
 class AddEmail(unittest.TestCase):

     @TmpConfig(["contact1.vcf", "contact2.vcf"])
```

If I do that (*temporarily*!) the other tests all run fine.


I noticed that the birthday field and phone numbers as well as ZIP codes are now wrapped in quotes. Do you know if there is a way to only quote strings that are problematic (like the @ at the beginning)?

File: khard/helpers/__init__.py [hunk 10]
```diff
@@ -61,6 +62,121 @@ def get_random_uid() -> str:
     return ''.join([random.choice(string.ascii_lowercase + string.digits)
                     for _ in range(36)])
 
+def yaml_clean(value: Optional[Union[str, List, Dict[str, Any], Sequence]]
+              ) -> Optional[Union[Sequence, List, str, Dict[str, Any],
```
[the remark this answers is not in the document]
Comment (skowalak) [reply khard/helpers/__init__.py:66]: I got a mypy error without the Sequence, meaning one of the properties of `YAMLEditable` that are used in `to_yaml` is of a Sequence type. I removed List though, thank you.

**Comment by skowalak** [comment]
@lucc I implemented the changes you requested.
About the quoting, I am not entirely sure. I believe ZIP codes at least are quoted because they are strings consisting of numbers only and ruamel quotes them so they are not interpreted as `tag:yaml.org,2002:int`, which would for example remove leading zeroes. That does not apply to birthdays (as they include dashes), still they are quoted, when according to the yaml spec they do not have be quoted. I had a look over at [ruamels source code](https://sourceforge.net/p/ruamel-yaml/code/ci/default/tree/emitter.py) but could not find out why.

So, if that is a big problem we could ofc ask Anthon van der Neut from ruamel if he knows if it is possible not to quote this.

**Review by lucc — APPROVED** [review]

**Comment by lucc** [comment]
thank you  @skowalak !

We can leave the quoting for later. That is just aesthetics in my opinion.  The important thing is that yaml editing should be more robust now that #276 should be fixed.

**Pull request merged**